### PR TITLE
Shebang

### PR DESCRIPTION
**Pull request by basarat**
Closes https://github.com/Microsoft/TypeScript/issues/2749

Appreciate a review by @CyrusNajmabadi : Emit ended up as a strategic call similar to detached comments

Also wrote : http://basarat.gitbooks.io/typescript/content/docs/compiler/contributing.html since I was in the area :rose:

### CODE REVIEW

File: tests/baselines/reference/shebang.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+//// [shebang.ts]
+#!/usr/bin/env node
+var foo = 'I wish the generated JS to be executed in node';
+
+
+//// [shebang.js]
+#!/usr/bin/env node
```
Comment (CyrusNajmabadi) [inline tests/baselines/reference/shebang.js]: So, does node ignore shebangs?

Comment (basarat) [inline tests/baselines/reference/shebang.js]: Yes.

Comment (CyrusNajmabadi) [inline tests/baselines/reference/shebang.js]: Can you add tests where the # isn't on the first position.  Thanks!

File: src/compiler/scanner.ts [hunk 2]
```diff
@@ -617,6 +641,16 @@ namespace ts {
     export function getTrailingCommentRanges(text: string, pos: number): CommentRange[] {
         return getCommentRanges(text, pos, /*trailing*/ true);
     }
+    
+    /** Optionally, get the shebang */
+    export function getShebang(text: string): string {
+        if (!shebangTriviaRegex.test(text)) {
```
Comment (CyrusNajmabadi) [inline src/compiler/scanner.ts]:
```
return shebangTriviaRegex.test(text)
    ? shebangTriviaRegex.exec(text)[0]
    : undefined;
```

File: src/compiler/scanner.ts [hunk 3]
```diff
@@ -1087,6 +1121,18 @@ namespace ts {
                     return token = SyntaxKind.EndOfFileToken;
                 }
                 let ch = text.charCodeAt(pos);
+
+                // Special handling for shebang
+                if (ch == CharacterCodes.hash &&  pos === 0 && isShebangTrivia(text, pos)) {
```
Comment (CyrusNajmabadi) [inline src/compiler/scanner.ts]: remove space after &&

**Comment by CyrusNajmabadi** [comment]
Looks pretty good to me.  With a couple of changes this should be good to go in.   I'd like @ahejlsberg to weigh in if we want to go ahead with them.  

My only concern is how this affects our ability to do something like add support for ```#pp-directives``` in the future.  Or maybe that wouldn't be a problem as we'd still just detect ```#!``` as a special directive.

**Comment by basarat** [comment]
Done: 

 * [x] add tests where the #! isn't on the first position
 * [x] CR based cleanup

File: src/compiler/scanner.ts [hunk 4]
```diff
@@ -617,6 +641,13 @@ namespace ts {
     export function getTrailingCommentRanges(text: string, pos: number): CommentRange[] {
         return getCommentRanges(text, pos, /*trailing*/ true);
     }
+    
+    /** Optionally, get the shebang */
+    export function getShebang(text: string): string {
+        return shebangTriviaRegex.test(text)
+                ? shebangTriviaRegex.exec(text)[0]
```
Comment (CyrusNajmabadi) [inline src/compiler/scanner.ts]: too much indent.

**Comment by CyrusNajmabadi** [comment]
@mhegazy You ok with this?

File: src/compiler/scanner.ts [hunk 5]
```diff
@@ -528,6 +538,20 @@ namespace ts {
         return pos;
     }
 
+    let shebangTriviaRegex = /^#!.*/;
```
Comment (mhegazy) [inline src/compiler/scanner.ts]: const?

File: tests/baselines/reference/shebang.js (hunk 1 above)
Comment (mhegazy) [inline tests/baselines/reference/shebang.js]: +1

Comment (basarat) [inline tests/baselines/reference/shebang.js]: Done :rose:

**Comment by basarat** [comment]
Done (as an amendment on the CR commit):

* [x] `const` for regex
* [x] further cleanup on code indentation

**Comment by mhegazy** [comment]
thanks!

**Pull request merged**